### ISSUE TITLE
CmdPeriod leaks SynthDefs

### ISSUE THREAD

**Issue by telephon** [supercollider/supercollider#4457]
## Environment

* SuperCollider version:63783ae
* Operating system:10.13.6

## Steps to reproduce

```supercollider


// run this, and watch the SynthDef count of the server
(
c = s.statusWatcher.numSynthDefs;
x = fork {
	var z;
	loop {
		z = { SinOsc.ar([1200, 1000, 800].choose, 0, 0.1) }.asSynthDef;
		z.send;
		s.sendBundle(1, ["/d_free", z.name]);
		0.02.wait;
		
	}
}
)

// stop with this line
(
fork {
	x.stop;
	2.wait;
	[s.statusWatcher.numSynthDefs, c, s.statusWatcher.numSynthDefs == c ].postln;
}
)



// run this again, and watch the SynthDef count of the server
(
c = s.statusWatcher.numSynthDefs;
x = fork {
	var z;
	loop {
		z = { SinOsc.ar([1200, 1000, 800].choose, 0, 0.1) }.asSynthDef;
		z.send;
		s.sendBundle(1, ["/d_free", z.name]);
		0.02.wait;
		
	}
}
)

// stop with cmd-period: there are SynthDefs which haven't been freed
(
fork {
	CmdPeriod.run;
	2.wait;
	[s.statusWatcher.numSynthDefs, c, s.statusWatcher.numSynthDefs == c ].postln; // => false
}
)

```

## Expected vs. actual behavior

Also with CmdPeriod, all SynthDefs should have been freed.

This is relevant in fixing #3344.

Maybe this is a server bug, because the messages have been sent off already and are waiting in the server queue.

**Comment by snappizz** [comment]
hi julian, i don't understand -- why should CmdPeriod free all SynthDefs?

**Comment by joshpar** [comment]
I agree - SynthDef creation / loading / storage and Command Period shouldn’t be tied. It never has, and a lot of things rely on this behavior (NRT for one).

This seems to me more like a problem with shortcuts being used for SynthDef creation - in this case, using {}.asSynthDef - if these need different treatment, that method should have its own pool of things to fire on CmdPeriod.

Josh

**Comment by jamshark70** [comment]
This seems to be a specific case: if /d_free messages have been sent and CmdPeriod intervenes, then the /d_free messages might not execute.

I think it's a reasonable expectation that all SynthDefs that you have explicitly requested to free should be freed.

I don't see anything in the issue description about CmdPeriod implicitly freeing SynthDefs in *all* cases -- despite the word choice at the end of the issue report, the context is the specific case of explicit usage of /d_free.

**Comment by telephon** [comment]
On `freeAll` (called in `CmdPeriod.run`), sclang sends the `/clearSched` command. This clears the pending `/d_free` message from the queue (as advertised). While this is good for things like `/s_new`, it isn't good for things like `/d_free`.

**Comment by telephon** [comment]
One workaround might be to never schedule such messages into the future, but have them done always immediately.

**Comment by jamshark70** [comment]
> One workaround might be to never schedule such messages into the future, but have them done always immediately.

That seems a bit dodgy. It seems possible that you could schedule a /s_new for, say, 1 seconds later and a /d_free for 1.5 seconds later. If /d_free is always immediate, then the /s_new would fail.

I'm rethinking this. Another argument would be that /clearSched should always do what it says, and it would be user code's responsibility to have a failsafe if needed.

I can see that you're drawing an analogy here to synth nodes: if you schedule /n_free, and cmd-. intervenes, the ending state is that the node is gone. By analogy, if you schedule /d_free, and cmd-. intervenes, the ending state is that the node is gone

**Comment by jamshark70** [comment]
Damn mobile interface.

I can see that you're drawing an analogy here to synth nodes: if you schedule /n_free, and cmd-. intervenes, the ending state is that the node is gone. By analogy, if you schedule /d_free, and cmd-. intervenes, the ending state that you're proposing is that the SynthDef is gone. I'm not sure the analogy always holds...? Is it conceivable that one might want to cancel a /d_free and leave the SynthDef intact?
